File: Structures/LinkedLists/Searching/TernarySearch.py
'''input
10
1
2
3
4
5
6
7
8
10
11
10
'''
size = int(input())

arr = []

for _ in range(size):
	element = int(input())
	arr.append(element)

value = int(input()) # value to be searched

lo = 0
hi = size - 1
print(arr)
flag = False
while lo <= hi:
	mid1 = lo + (hi-lo)//3
	mid2 = lo + 2*(hi-lo)//3

	if arr[mid1] == value:
		print("Value found at index {}".format(mid1))
		flag = True
		break
	if arr[mid2] == value:
		print("Value found at index {}".format(mid2))
		flag = True
		break

	if value < arr[mid1]:
		hi = mid1 - 1
	elif value > arr[mid1]:
		lo = mid2 + 1
	else:
		lo = mid1 + 1
		hi = mid2 - 1

if flag == False:
	print("Value Not Found")

File: Structures/LinkedLists/Searching/Ternary-Search.cpp
#include <bits/stdc++.h>
using namespace std;

int ternary_search(int array[],int l,int r, int x)
{
    if(r>=l)
    {
        int mid1 = l + (r-l)/3;
        int mid2 = r -  (r-l)/3;
        if(array[mid1] == x)
            return mid1;
        if(array[mid2] == x)
            return mid2;
        if(x<array[mid1])
            return ternary_search(array,l,mid1-1,x);
        else if(x>array[mid2])
            return ternary_search(array,mid2+1,r,x);
        else
            return ternary_search(array,mid1+1,mid2-1,x);

    }
    return -1;
}
int main() {

    int size;
    cin>>size;
    int array[size];
    for(int i=0;i<size;i++)
    {
        cin>>array[i];
    }
    int val;
    cin>>val;
    cout<<ternary_search(array,0,size-1,val)<<endl;
return 0;
}
